Simplify battle calc

DanVanAtta opened the pull request:
Main change is to accept all parameters at once with a builder object at the time a calculation is requested. This has a number of implications, notably we can remove much of the state tracking and the listener to notify when initial state had been created (initial state is now a no-op for the IOddsCalculator)
    
The ConcurrentOddsCalculator calls OddsCalculator, removing much of the tracking we can write that class more simply with java executor service.
    
TODO (needs some more testing):
    - we now pass a game data when calculate button is clicked
      - verify the copy of game data cases, can we actually mangle a game data? It does appear we only use it for read only data anyways and do not write back to the object
      - verify large game data cases. World at War around round 20 can be a 5MB file. The delay for the calculate button was very noticeable. At least now the button would be clickable right away, but we may see an extensive lag. If so we can look into ways to improve that performance.
     - fix checkstyle violations, run formatter, clean up imports

---------------

Goal here was to simplify, but the performance looked okay. I got no worse than current master, and on repeated runs speed increased by a factor of 5 (1.5s -> 0.3s, same setup as in: https://github.com/triplea-game/triplea/issues/2529#issuecomment-338856323)

RoiEXLab commented:
There are a couple of checkstyle violations causing the build to fail...

ron-murhammer commented:
@DanVanAtta You'll also need to verify this doesn't negatively impact AI performance since the Hard AI uses the battle calc extensively.

DanVanAtta commented:
@ron-murhammer AI looks good. Beyond that there is no reason to think it would be slower when OddsCalculator had stuff removed and runs as fast/faster.

@RoiEXLab Indeed, was noted as a todo, but should be fixed now.

-----------

What I'd really like reviewed at the moment while I'm still checking this out:
- if the lack of game data copies is really okay, seems so far like it should be, but a second set of :eyes:  on this would be good
- ConcurrentOddsCalculator was not tested, I did not create a new one, but it was rewritten. Extra scrutiny there is welcome as that is a rewrite and not just a refactor where a bunch of state tracking code was removed.


I'm still checking that the battle calc works well for large file sizes. Previously a delay of 5s or more could be observed before the calculate button was made available when launching the battle calculator screen. Now that delay is gone :)

ssoloff commented:
d4d9fb9 Renaming the `oddsCalculator` package to `oddscalculator` will break save game compatibility (see #2206).  Are we committed to the next release being backwards incompatible?

DanVanAtta commented:
To fix checkstyle violations effectively I had to reset the working set.  Working that back in as a new commit is not a valuable use of time, so this branch is just one squashed commit now.

I'm concerned we are too focused on checkstyle and eclipse compilation warnings and are missing much more important issues. I think we should focus reviews on high level first, see if there are any major changes needed, then save the checkstyle churn for last.

DanVanAtta commented:
I didn't mention, during that same squash I also moved most class files back to the original package, moved the newly split out inner classes along with a few UI classes to a new package 'oddscalc' and added TODO readme to merge that back with the badly named package

DanVanAtta commented:
> d4d9fb9 Renaming the oddsCalculator package to oddscalculator will break save game compatibility (see #2206). Are we committed to the next release being backwards incompatible?

Good call out, familiar story. No, we're not committed quite yet to a new backward compatible release.

RoiEXLab commented on src/main/java/games/strategy/triplea/oddsCalculator/ta/ConcurrentOddsCalculator.java:
```diff
@@ -1,437 +1,49 @@
+    final int totalRunCount = parameters.runCount;
+    for (int i = 0; i < totalRunCount && !cancelled; i++) {
+      parameters.setRunCount(1);
+      executor.submit(() -> aggregateResults.addResult(new OddsCalculator().doSimulation(parameters)));
```
Since `ExecutorService#submit` is not blocking, the cancelled boolean won't do much.
Solution to this would be to either make the Executor service a field, which is shared among all calls to calculate and to call `ExecutorService#shutdownNow` in the cancel method, or to create a Collection (LinkedList or Set, not sure which is better), which temporarily stores all instances of OddsCalculator and calls cancel on them (probably very resource intensive)

DanVanAtta replied:
Good catch, cancel was broken. dc6a183 fixes it.

RoiEXLab commented on src/main/java/games/strategy/triplea/oddsCalculator/ta/ConcurrentOddsCalculator.java:
```diff
@@ -1,437 +1,49 @@
-      // there is a small chance calculate or setCalculateData or something could be called in between calls to
-      // setGameData
-      latchSetData.await();
+      executor.awaitTermination(15, TimeUnit.SECONDS);
```
Is there any reasoning why this exact timeout has been chosen?
I'd suggest using using the boolean value returned by this method somewhere, ~to indicate the awaitTermination method has exceeded the timeout when it does~ to print a warning or something similar, that this method is taking way too long to execute

DanVanAtta replied:
a676536 adds some reasoning. Essentially 15s is beyond a very long time to wait for a calc, most are down in 2 to 5 seconds.

The error handling could be yet better, but at least we are timing out.

RoiEXLab replied:
I'd suggest using an ExecutorCompletionService here, which essentially wraps an executor service and allows you to retrieve future objects, and allows for easier error handling.
Essentially what I did in #2558

DanVanAtta replied:
I don't think we would want to even show an error dialog if we hit the timeout. The handling for now would be to cut the calc and display a result. If there were any results at all, there would be some stats. If we had error handling to show an error, ideally it would be done in background via network request to server for logging.

We could instead show a big error message to a player while playing the game. If they get zero results they'll likely file a bug report or move on. An error message even though there are results is more likely to cause the player confusion, eg: they may wonder if they should restart the game.

Given those scenarios, just cutting the calc at 15 seconds and not doing any further error handling is a reasonable scope and an improvement over the previous potential infinite wait. Hence I'm not inclined to update the error handling further for this scenario.

RoiEXLab commented on src/main/java/games/strategy/triplea/oddsCalculator/ta/OddsCalculator.java:
```diff
@@ -5,272 +5,101 @@
-  public static final String OOL_ALL_REGEX = "\\*";
   public static final String OOL_SEPARATOR = ";";
-  public static final String OOL_SEPARATOR_REGEX = ";";
+  private static final String OOL_SEPARATOR_REGEX = ";";
```
I don't think 2 constants for the exact same final string are needed

DanVanAtta replied:
I would tend to agree. Though I'm only fixing easy compiler warnings here. Beyond stated scope, I'd rather not expend effort to determine original intent nor complicate the diff further with items that are more than removing unnecessary tracking code revolving around GameData.

RoiEXLab commented on src/main/java/games/strategy/triplea/oddsCalculator/ta/OddsCalculator.java:
```diff
@@ -5,272 +5,101 @@
-  static boolean isValidOrderOfLoss(final String orderOfLoss, final GameData data) {
+  /**
+   * Validates if an order of loss string is good or not.
+   * TODO: refactor this method... consider a better javaodc that tell us what a valid OOL string would be.
```
Nit: Spelling javaodc -> javadoc

RoiEXLab commented on src/main/java/games/strategy/triplea/oddscalc/OddsCalculatorPanel.java:
```diff
@@ -352,16 +319,14 @@ private void updateStats() {
     final AtomicReference<AggregateResults> results = new AtomicReference<>();
     final WaitDialog dialog =
-        new WaitDialog(this, "Calculating Odds (" + calculator.getThreadCount() + " threads)", new AbstractAction() {
+        new WaitDialog(this, "Calculating Odds", new AbstractAction() {
```
Note to myself: When modifying the WaitDialog Constructor to accept an ActionListener instead of an Action, this abstract action can be replaced with a lambda expression
Actually, the only public usage of the second constructor is this line, so maybe within the scope of this PR, I'll let you decide

Nit: This line now can now bemerged with the line above

DanVanAtta replied:
The update here was to change the message, I want to keep the diff and any potential rollbacks to exactly that.

DanVanAtta replied:
Sorry, I don't quite follow on the line merge.
If this were refactored, I would suggest merging wait panel and dialog with JDialogBuilder so we could write something like:
```
JDialogBuilder.builder()
  .title("CalculatingOdds")
  .blocking(true)
  .cancelAction(() -> cancelAction())
  .showDialog();
```

RoiEXLab replied:
Line merge just meant this line now fits in the same line as the declaration.

Your suggestion would work too as well, but would require far more work, and is clearly out-of-scope, while changing a param type to an interface type is far easier to do

DanVanAtta commented on src/main/java/games/strategy/performance/PerfTimer.java:
```diff
@@ -56,7 +61,8 @@ static boolean isEnabled() {
 
-  static PerfTimer startTimer(final String title) {
+  @SuppressWarnings("unused")
+  public static PerfTimer startTimer(final String title) {
```
Intellij default. The method is actually unused, the warning flag and @suppresswarnings both are useful to avoid removal.

DanVanAtta commented:
Locally the build does passes, though Travis seems to be having issues at the moment,
```
An error occurred while generating the build script.
```

DanVanAtta commented on src/main/java/games/strategy/triplea/oddscalc/DummyPlayer.java:
```diff
@@ -0,0 +1,230 @@
+/**
+ * Do-nothing implementation to support battle calc.
+ */
+public class DummyPlayer extends AbstractAI {
```
This code is all just moved, it was from an inner class, same with any of the other Dummy* classes.

DanVanAtta commented:
Updated, needs review @ron-murhammer  @RoiEXLab

ssoloff commented on src/main/java/games/strategy/triplea/oddsCalculator/ta/ConcurrentOddsCalculator.java:
```diff
@@ -1,437 +1,58 @@
-  private static final Logger logger = Logger.getLogger(ConcurrentOddsCalculator.class.getName());
-  private static final int MAX_THREADS = Math.max(1, Runtime.getRuntime().availableProcessors());
+public class ConcurrentOddsCalculator extends OddsCalculator {
+  private final Collection<OddsCalculator> workers = new HashSet<>();
```
This field is apparently unused.

DanVanAtta replied:
2bd0208

ssoloff commented on src/main/java/games/strategy/triplea/oddsCalculator/ta/ConcurrentOddsCalculator.java:
```diff
@@ -1,437 +1,58 @@
-      }
-    }
-  }
+  private boolean cancelled = false;
```
This field should probably be `volatile` because it is read/written from/to by different threads.

DanVanAtta replied:
cde48b2

ssoloff commented:
While testing this branch, I observed the following exception when I attempted to close the Odds Calculator window via the system close icon (the "X" in the upper left/right corner, depending on your OS):

```
java.lang.NullPointerException
	at games.strategy.triplea.oddscalc.OddsCalculatorPanel.shutdown(OddsCalculatorPanel.java:183)
	at games.strategy.triplea.oddscalc.OddsCalculatorDialog.dispose(OddsCalculatorDialog.java:70)
	at javax.swing.JDialog.processWindowEvent(JDialog.java:691)
	at java.awt.Window.processEvent(Window.java:2013)
	at java.awt.Component.dispatchEventImpl(Component.java:4889)
	at java.awt.Container.dispatchEventImpl(Container.java:2294)
	at java.awt.Window.dispatchEventImpl(Window.java:2746)
	at java.awt.Component.dispatchEvent(Component.java:4711)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:758)
	at java.awt.EventQueue.access$500(EventQueue.java:97)
	at java.awt.EventQueue$3.run(EventQueue.java:709)
	at java.awt.EventQueue$3.run(EventQueue.java:703)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:80)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:90)
	at java.awt.EventQueue$4.run(EventQueue.java:731)
	at java.awt.EventQueue$4.run(EventQueue.java:729)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:80)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:728)
	at games.strategy.engine.framework.GameRunner$1.dispatchEvent(GameRunner.java:330)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:201)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
```

ssoloff commented on src/main/java/games/strategy/triplea/oddscalc/OddsCalculatorPanel.java:
```diff
@@ -256,25 +241,6 @@ private void setupListeners() {
-      public void mouseMoved(final MouseEvent e) {
-        final String memoryAvailable = "<br/>Percentage of memory available: "
-            + String.format("%.2f", (percentageOfFreeMemoryAvailable() * 100)) + "% <br/>Free memory available: "
-            + (freeMemoryAvailable() / (1024 * 1024)) + "MB <br/>Maximum allowed memory: "
```
The `freeMemoryAvailable()` method in this class is no longer used and can be removed.

DanVanAtta replied:
87ab122

ssoloff commented on src/main/java/games/strategy/triplea/oddsCalculator/ta/ConcurrentOddsCalculator.java:
```diff
@@ -1,437 +1,58 @@
+    for (int i = 0; i < totalRunCount; i++) {
+      executor.submit(() -> {
+        if (!cancelled) {
+          aggregateResults.addResult(OddsCalculator.doSimulation(parameters));
```
Multiple threads are calling the `AggregateResults#addResult()` method, but access to the corresponding `AggregateResults#m_results` field does not appear to be thread safe (there are no locks and the field is of the thread-unsafe type `ArrayList`).

DanVanAtta replied:
4960d79

DanVanAtta commented:
Good test case @ssoloff , fixed in: dd4b85e

DanVanAtta commented:
Good feedback @ssoloff , should all be addressed in the noted commits, please let me know what you think.

ssoloff commented on src/main/java/games/strategy/triplea/oddsCalculator/ta/AggregateResults.java:
```diff
@@ -13,14 +14,17 @@
 
   public AggregateResults(final int expectedCount) {
-    m_results = new ArrayList<>(expectedCount);
+    m_results = Collections.synchronizedList(new ArrayList<>(expectedCount));
```
I _think_ this is sufficient to ensure there are no thread safety issues in `AggregateResults`.  My brief analysis suggests that once `m_results` is populated by the calculation thread(s), it is never again modified, so that the various `getXXX()` methods in this class do not need to worry about the state of `m_results` possibly changing while they are executing (if it could change, even while synchronized, many of the `getXXX()` methods would then not be atomic, and could yield incorrect results).

If it's true that `m_results` is never again modified once the simulation runs are complete, `AggregateResults` could be made immutable and the thread safety discussion would be moot.  What if each task created by `ConcurrentOddsCalculator` simply pushed its `BattleResults` to a thread-safe collection that was local to `calculate()`?  Then when all tasks have finished, a new `AggregateResults` could be created from that collection with an immutable non-thread-safe copy?  The following incomplete diff captures what I'm thinking:

```diff
diff --git a/src/main/java/games/strategy/triplea/oddsCalculator/ta/AggregateResults.java b/src/main/java/games/strategy/triplea/oddsCalculator/ta/AggregateResults.java
index 03d0cc1..6df890d 100644
--- a/src/main/java/games/strategy/triplea/oddsCalculator/ta/AggregateResults.java
+++ b/src/main/java/games/strategy/triplea/oddsCalculator/ta/AggregateResults.java
@@ -23,6 +23,11 @@ public class AggregateResults implements Serializable {
   private final List<BattleResults> m_results;
   private long m_time;
 
+  public AggregateResults(final long time, final Collection<BattleResults> results) {
+    m_time = time;
+    m_results = new ArrayList<>(results);
+  }
+
   public AggregateResults(final int expectedCount) {
     m_results = Collections.synchronizedList(new ArrayList<>(expectedCount));
   }
diff --git a/src/main/java/games/strategy/triplea/oddsCalculator/ta/ConcurrentOddsCalculator.java b/src/main/java/games/strategy/triplea/oddsCalculator/ta/ConcurrentOddsCalculator.java
index cdba242..861d329 100644
--- a/src/main/java/games/strategy/triplea/oddsCalculator/ta/ConcurrentOddsCalculator.java
+++ b/src/main/java/games/strategy/triplea/oddsCalculator/ta/ConcurrentOddsCalculator.java
@@ -21,14 +23,14 @@ public class ConcurrentOddsCalculator extends OddsCalculator {
   public AggregateResults calculate(final OddsCalculatorParameters parameters) {
     final ExecutorService executor = Executors.newFixedThreadPool(Runtime.getRuntime().availableProcessors());
     final long start = System.currentTimeMillis();
-    final AggregateResults aggregateResults = new AggregateResults(parameters.runCount);
+    final Queue<BattleResults> battleResults = new ConcurrentLinkedQueue<>();
 
     final int totalRunCount = parameters.runCount;
     parameters.setRunCount(1);
     for (int i = 0; i < totalRunCount; i++) {
       executor.submit(() -> {
         if (!cancelled) {
-          aggregateResults.addResult(OddsCalculator.doSimulation(parameters));
+          battleResults.add(OddsCalculator.doSimulation(parameters));
         }
       });
     }
@@ -40,8 +42,8 @@ public class ConcurrentOddsCalculator extends OddsCalculator {
     } catch (final InterruptedException e) {
       Thread.currentThread().interrupt();
     }
-    aggregateResults.setTime(System.currentTimeMillis() - start);
-    return aggregateResults;
+
+    return new AggregateResults(System.currentTimeMillis() - start, battleResults);
   }
 
   /*
```

It's kind of six of one, half-dozen of another.  I'm not sure it's worth it, but it does keep all the threading concerns local to `ConcurrentOddsCalculator`.

<hr>

**NOTE:** Another thing that needs to be considered with the use of `Collections#synchronizedList()` is that if an instance of `AggregateResults` is deserialized from an older version, the resulting `m_results` will not be synchronized.  Not sure if that's a problem.  I'm assuming a deserialized `AggregateResults` would never be accessed concurrently.

<hr>

**NOTE:** It appears the following methods in this class are now unused:

* `addResults()`
* `getResults()`

ssoloff replied:
@DanVanAtta I submitted a PR to your branch that more completely fleshes-out the `AggregateResults` immutability proposal for your consideration.

DanVanAtta commented:
Seems that all feedback is addressed, dev testing has been good. Let's get this merged and out to the testers. There is a big performance improvement here where there is no delay for the battle calculator button to be active. Most games past round 10 or 15 will have a multi-second delay before battle calc can be played (due to GameData copy that is now removed).

ssoloff commented on src/main/java/games/strategy/triplea/ai/proAI/ProAI.java:
```diff
@@ -199,7 +196,6 @@ protected void purchase(final boolean purchaseForBid, final int pusToSpend, fina
       } finally {
         data.releaseReadLock();
       }
-      calc.setData(dataCopy);
```
@ron-murhammer Kind of related to #2600...  In the current implementation, it doesn't look like `dataCopy` will be used by the Battle Calculator anymore.  Even though it's passed to `simulateBattles()`, it never gets passed to the calculator, which ends up just using the original game data.

Just want to confirm if not using this possibly-different copy of the `GameData` is a problem that needs to be fixed, as well.

ron-murhammer replied:
The AI needs to pass a copy of game data when it simulates move and battles before purchasing. So either needs to set a copy of the data or create a new battle calc with the data. This is definitely not ok to have removed.

ssoloff replied:
Ok, I'll see how difficult that is to work into the new architecture.  A complete revert is looking more attractive, though. 😞

ron-murhammer replied:
@ssoloff Yeah. I'm fine with a full revert on this if that is the easiest path forward.
